qiskit/transpiler/target.py

```diff
@@ -0,0 +1,600 @@
+                    out_durations.append((instruction, list(qarg), properties.length, "s"))
+        self._instruction_durations = InstructionDurations(out_durations)
+        return self._instruction_durations
+
+    def timing_constraints(self):
+        """Get an :class:`~qiskit.transpiler.TimingConstraints` object from the target
+
+        Returns:
+            TimingConstraints: The timing constraints represented in the Target
+        """
+        return TimingConstraints(
+            self.granularity, self.min_length, self.pulse_alignment, self.aquire_alignment
+        )
+
+    def instruction_schedule_map(self):
+        """Return an :class:`~qiskit.pulse.InstructionScheduleMap` for the
+        instructions in the target with a pulse schedule defined.
+
+        Returns:
+            InstructionScheduleMap: The instruction schedule map for the
+            instructions in this target with a pulse schedule defined.
+        """
+        if self._instruction_schedule_map is not None:
+            return self._instruction_schedule_map
+        out_inst_schedule_map = InstructionScheduleMap()
+        for instruction, qargs in self._gate_map.items():
+            for qarg, properties in qargs.items():
+                if properties is not None and properties.schedule is not None:
+                    out_inst_schedule_map.add(instruction, qarg, properties.schedule)
+        self._instruction_schedule_map = out_inst_schedule_map
+        return out_inst_schedule_map
+
+    def get_instruction_from_name(self, instruction):
+        """Get the :class:`~qiskit.circuit.Instruction` object for a given name
+
+        Args:
+            instruction (str): The instruction name to get the
+                :class:`~qiskit.circuit.Instruction` instance for
+        Returns:
+            Instruction: The Instruction instance corresponding to the name
+        """
+        return self._gate_name_map[instruction]
+
+    def get_instructions_for_qargs(self, qarg):
+        """Get the qargs for a given gate by name
+
+        Args:
+            qarg (tuple): A qarg tuple of the qubits to get the gates that apply
+                to it. For example, ``(0,)`` will return the set of all
+                instructions that apply to qubit 0.
+        Returns:
+            list: The set of :class:`~qiskit.circuit.Instruction` instances
+            that apply to the specified qarg.
+
+        Raises:
+            KeyError: If qarg is not in target
+        """
+        if qarg not in self._qarg_gate_map:
+            raise KeyError(f"{qarg} not in target.")
+        return [self._gate_name_map[x] for x in self._qarg_gate_map[qarg]]
+
+    @property
+    def instruction_names(self):
+        """Get the instruction names in the target."""
+        return set(self._gate_map)
+
+    @property
+    def instructions(self):
+        """Get the :class:`~qiskit.circuit.Instruction` objects in the target."""
+        return list(self._gate_name_map.values())
```

Comment:
Right now this returns something like:
```py
[Instruction(name='rx', num_qubits=1, num_clbits=0, params=[Parameter(theta)]),
 Instruction(name='rx', num_qubits=1, num_clbits=0, params=[0.5235987755982988]),
 Instruction(name='u', num_qubits=1, num_clbits=0, params=[Parameter(theta), Parameter(phi), Parameter(lambda)]),
 Instruction(name='cx', num_qubits=2, num_clbits=0, params=[]),
 Instruction(name='measure', num_qubits=1, num_clbits=1, params=[]),
 Instruction(name='ecr', num_qubits=2, num_clbits=0, params=[])]
```

But isn't this more simply expressed as something like:
```py
[RXGate(Parameter(theta)),
RXGate(0.5235987755982988),
UGate(Parameter(theta), Parameter(phi), Parameter(lambda)),
CXGate(),
Measure(),
ECRGate()
]

This naturally links the gates to their stdlib implementation (which knows how many qubits it acts on etc.).. do you store that somewhere else right now?
```

Reply:
The return from `target.instructions` should be this already. The `Instruction`/`Gate` object that gets added in `add_instruction` is what is being returned here. 

I think what you're seeing is just an artifact of the `__repr__` for `Instruction`/`Gate`. Mainly that if you do `print(RXGate(3.14))` it prints `Instruction(name='rx', num_qubits=1, num_clbits=0, params=[3.14])`. When I do `print(type(FakeBackendV2().target.instructions[0]))` I get: `<class 'qiskit.circuit.library.standard_gates.rx.RXGate'>`

Reply:
Oh ok. Makes sense that is just the __repr__ I'm seeing. 

But this raises another question: in an earlier comment Paul brought up the issue of two different kinds of Measure being available (say ESP readout and normal readout) which could have different errors and durations. A similar example is when you calibrate two kinds of CNOTs on the same qubits, say a slow_cx and a fast_cx. How would this be supported? If a `Gate/Instruction` object is stored in this datastructure, would it e.g. store something like this?
- `CXGate(name='fast_cx')` & `CXGate(name=slow_cx)`
- `Measure(name='measure')` & `Measure(name='esp_measure')`

If this is supported it would be good to add it to FakeBackendV2 as an example. I know the transpiler will have difficulty dealing with this since it looks for gate name, but at least we can make the Backend support such a thing and then we make the transpiler work with gate objects instead.

Reply:
It should work today in a similar manner to how I did `rx_30` and `rx` in the examples. This is explicitly a use case I was thinking about and one @nkanazawa1989 and I talked about at some length for calibration experiments so a user can calibrate custom variants of a gate and have the transpiler use that. In the case of cx gate you would do something like:

```python
target = Target()
target.add_instruction(CXGate(), {(0, 1): InstructionProperties(length=1e-7, 1e-3), (1, 2): InstructionProperties(length=1e-7, 1e-3)})
target.add_instruction(CXGate(), {(0, 2): InstructionProperties(length=1e-6, 1e-2)}, name='cx_slow')
```

Similarly for measure and esp_readout I left a comment on this a while ago here: https://github.com/Qiskit/qiskit-terra/pull/5885#discussion_r712322791
but it would be something like:

```python
     measure_props = {
            (0,): GateProperties(length=4e-8, error=0.0002),
            (1,): GateProperties(length=4e-8, error=0.0002),
        }
        gmap.add_gate(Measure(),  [(0,), (1,)], properties=measure_props)
        
        esp_readout_props = {
            (0,): GateProperties(length=2e-8, error=0.0001),
            (1,): GateProperties(length=2e-8, error=0.00012),
        }
        # esp readout behaves like a normal measure but has a unique name to distinguish it from a normal measurement
        gmap.add_gate(Measure(),  [(0,), (1,)], properties=esp_readout_props, name="esp_readout")
```
As for the transpiler component yeah there is a bunch of work there to make the transpiler pick between multiple implementations. But at least with how I implemented the basis translator PR in #7227 if the operations are on different arguments it's not a ton of extra work (they just have to get the name->gate relationship and use that for the equivalence library). For gates on the same arguments though it's trickier because you need a way to pick between them.